### DISCUSSION TITLE
Tailwind CSS load order with Element Plus in Vite and Vue

### DISCUSSION THREAD

**Discussion by vildanbina** [tailwindlabs/tailwindcss#11764] [Help]
Hello,

I am currently using Tailwind CSS with Element Plus in a Vite and Vue application. I am importing Element Plus on demand. However, I am facing an issue where the styles from Tailwind CSS are getting overridden by Element Plus because Tailwind CSS is loaded after Element Plus.

This is causing some inconsistency in the styles in my application. I would like to find a way to make Tailwind CSS load before Element Plus and all other styles to avoid these overwriting issues.

Here's a simplified version of my setup:

##### vite.config.ts:
```ts
export default defineConfig({
    resolve: {
        alias: [{ find: "@", replacement: resolve(__dirname, "src/") }],
    },
    plugins: [
        vue(),
        ElementPlus({
            useSource: true,
        }),
        AutoImport({
            resolvers: [ElementPlusResolver({ importStyle: "sass" })],
        }),
        Components({
            resolvers: [ElementPlusResolver({ importStyle: "sass" })],
        }),
    ],
});
```

##### main.ts:
```ts
import { createApp } from "vue";
import "./assets/scss/tailwind.scss";
import App from "./App.vue";

const app = createApp(App);

app.mount("#app");
```

If there is any configuration or step I am missing to make Tailwind CSS load before Element Plus and all other styles, I would greatly appreciate your guidance.

Thank you.

**Comment by ch00z00** [comment]
Apologies for the confusion. It seems like the styles are still being loaded in an order that's causing conflicts. To ensure that Tailwind CSS is loaded before Element Plus styles, you can consider the following steps:

1. **Use `vite-plugin-windicss`**:
   The `vite-plugin-windicss` plugin provides integration for Windi CSS, which is a utility-first CSS framework like Tailwind CSS, but it's specifically designed to work well with Vite. Windi CSS automatically scans your HTML and Vue files for used classes and generates optimized styles.

   To set up `vite-plugin-windicss`, you can follow these steps:

   Install the plugin:
   ```
   npm install vite-plugin-windicss --save-dev
   ```

   Modify your `vite.config.js` to include the plugin and configure it:
   ```javascript
   import WindiCSS from 'vite-plugin-windicss';
   
   export default {
     // ...
     plugins: [
       // ...
       WindiCSS(),
     ],
   };
   ```

   Remove any explicit imports of Tailwind CSS in your code (like `import "./assets/scss/tailwind.scss";`) as Windi CSS will automatically handle this.

2. **Remove Direct Element Plus Imports**:
   If you're using `vite-plugin-auto-import` and `vite-plugin-components` to automatically import Element Plus, you might need to remove the manual Element Plus import in your Vue components. Let the auto-import plugins handle the imports for you.

3. **Order of Plugins**:
   Ensure that the `vite-plugin-windicss` plugin is included before other style-related plugins in the `plugins` array in your `vite.config.js`. This way, Windi CSS will process the styles first, which should help in avoiding conflicts.

Here's how your `vite.config.js` could look:

```javascript
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';
import WindiCSS from 'vite-plugin-windicss';
import AutoImport from 'vite-plugin-auto-import';
import ElementPlusResolver from 'unplugin-element-plus/vite';

export default defineConfig({
  plugins: [
    vue(),
    WindiCSS(),
    AutoImport({
      resolvers: [ElementPlusResolver({ importStyle: 'sass' })],
    }),
  ],
});
```

Remember to customize the Windi CSS configuration as needed to match your project's requirements.

By using `vite-plugin-windicss`, you should be able to avoid conflicts between Tailwind CSS and Element Plus styles while ensuring that the styles are loaded in the desired order.

**Reply by vildanbina** [reply]
I wanted to use Tailwind CSS not to switch on other CSS framework

**Reply by ch00z00** [reply]
In that case, we can adjust the order of styles loading within your current setup to ensure that Tailwind CSS is loaded before Element Plus styles. Here's how you can achieve this:

1. **Update `main.ts`**:
    
    In your `main.ts` file, you can import the Tailwind CSS first and then the Element Plus styles. This way, Tailwind CSS styles will be applied before Element Plus styles, potentially avoiding conflicts.
    
    ```jsx
    import { createApp } from "vue";
    import "./assets/scss/tailwind.scss"; // Import Tailwind CSS first
    import "element-plus/lib/theme-chalk/index.css"; // Then import Element Plus styles
    import App from "./App.vue";
    
    const app = createApp(App);
    app.mount("#app");
    
    ```
    
2. **Modify the Order of Plugins**:
    
    You can also try reordering the plugins in your `vite.config.ts` file to ensure that the styles from Tailwind CSS are processed before the Element Plus styles. Make sure the Windi CSS (or Tailwind CSS if you decide to stick with it) plugin comes before the Element Plus plugin:
    
    ```jsx
    export default defineConfig({
        plugins: [
            vue(),
            WindiCSS(), // or Tailwind CSS plugin
            ElementPlus({
                useSource: true,
            }),
            AutoImport({
                resolvers: [ElementPlusResolver({ importStyle: "sass" })],
            }),
            Components({
                resolvers: [ElementPlusResolver({ importStyle: "sass" })],
            }),
        ],
    });
    
    ```
    

By following these steps, you should be able to maintain your use of Tailwind CSS while ensuring that its styles are applied before Element Plus styles, potentially resolving the inconsistencies you've been facing.

**Reply by vildanbina** [reply]
@ch00z00, I'm considering switching to Windi, but I've noticed that Windi currently incorporates all the features from Tailwind 2. Have there been any updates to ensure compatibility with the features introduced in Tailwind 3?

**Reply by ch00z00** [reply]
@vildanbina 
We are not sure how compatible Windi is with Tailwind v3, as it is currently only fully compatible with v2. We are in the process of asking questions in Windi's Discussions, so please be patient!

<img width="661" alt="image" src="https://github.com/tailwindlabs/tailwindcss/assets/131364681/f4947011-6149-4876-b14b-9fd52850cec6">

(👆 Refer to from vite-plugin-windicss Features)

**Reply by vildanbina** [reply]
Thanks, please let me know if there is any updates

**Reply by ch00z00** [reply]
@vildanbina 
As I said, I could not find sufficient sources that Windi is fully compatible with Tailwind CSS v3.
Also, a question in Windi's Discussions did not yield a strong answer.

If you are considering using Vite and a CSS framework compatible with Tailwind CSS, [UnoCSS](https://unocss.dev/) may be more appropriate than Windi.

UnoCSS is an Atomic CSS engine compatible with Tailwind CSS.
It was started by Anthony Fu, a core member of Vite and Vue, and is currently under active development by several members.
Before developing UnoCSS, Anthony Fu had developed a similar Atomic CSS engine, Windi CSS, which was designed to be very strictly compatible with Tailwind, including the plugin mechanism, but the UnoCSS was redesigned from scratch to focus more on speed and flexibility of configuration, ignoring the detailed compatibility with Tailwind.

[This](https://unocss.dev/integrations/vite) is the Vite plugin.

**Comment by Red-Asuka** [comment]
I am experiencing the same issue. Does anyone have a better solution? I prefer using Tailwind over Windi.
